### PR TITLE
Fix the error when use spread arguments twice

### LINKED ISSUE #32835: Error when use spread arguments twice
**TypeScript Version:** 3.5.1

**Search Terms:**

**Code**

```ts
declare function fn (a: string, b: string, ...c: string[]): void
declare const foo: [string, string, ...string[]]
declare const bar: string[]
fn(...foo) // ok
fn(...foo, ...bar) // error
fn(...foo, ...foo) // error
fn(...bar, ...foo) // error
```

**Expected behavior:**

No errors

**Actual behavior:**

```
Expected at least 2 arguments, but got 1 or more
```

**Playground Link:** [link](https://www.typescriptlang.org/play/?noImplicitThis=false#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVPwAooAueAZwxi1QHMAaeAIzMursYDpuxWqbaAbQC6ASjIA3HFmAAoUJFgIweSkhw4ygtgMY6O8bp31Dhw+eGhx4K1Gqaw+7U7MSpCRxBtHwA9L-gcAGtXd08NLm4HGB9-eBAYGBwYUI9uLxxIzgzYgISklLc0zmisnL88xOSgA)

**Related Issues:**

### PR DESCRIPTION
**Pull request by rchaser53**
This is a partial fix for #32835, but a full fix should also handle the additional test posted there.

### CODE REVIEW

**Comment by rchaser53** [comment]
@RyanCavanaugh
Could you review my PR? And if it's ok, please merge it.
This bug is still caused using the current version.

CI is failed, but it doesn't seem to be my fault.
The other PRs are also failed and emit the same errors.
ex. https://github.com/microsoft/TypeScript/pull/36148/checks?check_run_id=385338469

**Comment by elibarzilay** [comment]
(@rchaser53 -- JFYI, I pushed a rebase to drop the big merge commit.)

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -24365,9 +24365,24 @@ namespace ts {
...
-                if (spreadArgIndex >= 0) {
-                    return spreadArgIndex >= getMinArgumentCount(signature) && (hasEffectiveRestParameter(signature) || spreadArgIndex < getParameterCount(signature));
+                const firstSpreadArgIndex = getSpreadArgumentIndex(args);
+                const existSpreadArgIndex = firstSpreadArgIndex >= 0;
+                if (existSpreadArgIndex) {
+                      if (firstSpreadArgIndex === args.length - 1) {
+                          return firstSpreadArgIndex >= getMinArgumentCount(signature) && (hasEffectiveRestParameter(signature) || firstSpreadArgIndex < getParameterCount(signature));
+                      }
+
+                      const totalCount = node.arguments.reduce((total, next) => {
```
Comment (elibarzilay) [inline src/compiler/checker.ts]: About using `reduce`: first, there is a preference for plain loops in the code, and there's another minor issue: you're using `getSpreadArgumentIndex` which will scan the list and test members with `isSpreadArgument`, and then the `reduce` tests again for `isSpreadArgument`.  If you switch to a plain loop, you can skip the re-test for all of these elements.

File: tests/cases/conformance/expressions/functionCalls/callWithSpread3.ts [hunk 2]
```diff
@@ -1,10 +1,20 @@
...
 takeTwo(...t2, 'a'); // error on 'a'
-takeTwo(...t3); // error on ...t3
\ No newline at end of file
+takeTwo(...t3); // error on ...t3
+
+// ok
+takeTwoOrMore(...t4);
+takeTwoOrMore(...t4, ...t5);
+takeTwoOrMore(...t4, ...t4);
+takeTwoOrMore(...t5, ...t4);
```
Comment (elibarzilay) [inline tests/cases/conformance/expressions/functionCalls/callWithSpread3.ts]: There should likely be a continuation issue for typechecking which still seems broken.  Just a quick example, this:

    declare const s2: [string, string]
    declare const n_: number[]
    takeTwoOrMore(...s2, ...n_);

fails with

    error TS2345: Argument of type 'number' is not assignable to parameter of type 'string'.

which is probably bogus due to code that does not handle spreaded tuples properly.

(But it should be fine to finish this PR without handling it and do this via a separate issue.)

**Comment by elibarzilay** [comment]
Note that I changed the PR text to keep #32835 for the additional issue there.

File: src/compiler/checker.ts (hunk 1 above)
Comment (rchaser53) [reply src/compiler/checker.ts]: I changed reduce to a plain loop. And I tried to remove Spread Argument but it failed. I think getSpreadArgumentIndex only searches the first spread element. So it needs to use isSpreadArgument for the case using two or more spread elements.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -24365,9 +24365,24 @@ namespace ts {
...
                 // If a spread argument is present, check that it corresponds to a rest parameter or at least that it's in the valid range.
-                const spreadArgIndex = getSpreadArgumentIndex(args);
-                if (spreadArgIndex >= 0) {
-                    return spreadArgIndex >= getMinArgumentCount(signature) && (hasEffectiveRestParameter(signature) || spreadArgIndex < getParameterCount(signature));
+                const firstSpreadArgIndex = getSpreadArgumentIndex(args);
+                const existSpreadArgIndex = firstSpreadArgIndex >= 0;
+                if (existSpreadArgIndex) {
+                      if (firstSpreadArgIndex === args.length - 1) {
+                          return firstSpreadArgIndex >= getMinArgumentCount(signature) && (hasEffectiveRestParameter(signature) || firstSpreadArgIndex < getParameterCount(signature));
+                      }
```
[the remark this answers is not in the document]
Comment (rchaser53) [reply src/compiler/checker.ts]: Oh... I didn't notice it. Honestly, I don't have the confidence to fix it correctly now. What should I do?

**Comment by rchaser53** [comment]
Thank you for the review. I fixed it as possible as I can.

**Comment by sandersn** [comment]
@rchaser53 Do you want to keep working on this? Let us know if we can help explain anything more.

**Comment by rchaser53** [comment]
I cannot have the time to resolve this issue now. My main work is too busy. I'm really sorry for taking your time... I close this PR.

**Pull request closed without merging**